python/cudf/cudf/core/buffer/utils.py

```diff
@@ -134,3 +140,92 @@ def get_spill_lock() -> Union[SpillLock, None]:
     _id = threading.get_ident()
     spill_lock, _ = _thread_spill_locks.get(_id, (None, 0))
     return spill_lock
+
+
+def _clear_property_cache(
+    instance_ref: weakref.ReferenceType[T], nbytes: int, attrname: str
+) -> Optional[int]:
+    """Spill handler that clears the `cached_property` of an instance
+
+    The signature of this function is compatible with SpillManager's
+    register_spill_handler.
+
+    To avoid keeping instance alive, we take a weak reference of the instance.
+
+    Parameters
+    ----------
+    instance_ref
+        Weakref of the instance
+    nbytes : int
+        Size of the cached data
+    attrname : str
+        Name of the cached attribute
+
+    Return
+    ------
+    int
+        Number of bytes cleared
+    """
+
+    instance = instance_ref()
+    if instance is None:
+        return 0
+
+    cached = instance.__dict__.get(attrname, None)
+    if cached is None:
+        return None  # The cached has been cleared
+
+    # If `cached` is known outside of the cache, we cannot free any
+    # memory by clearing the cache. We know of three references:
+    # `instance.__dict__`, `cached`, and `sys.getrefcount`.
+    if sys.getrefcount(cached) > 3:
+        return None
+
+    instance.__dict__.pop(attrname, None)  # Clear cache atomically
+    return nbytes
+
+
+class cached_property(functools.cached_property):
+    """A version of `cached_property` that delete instead of spill the cache
+
+    When spilling is disabled (the default case), this decorator is identical
+    to `functools.cached_property`.
+
+    When spilling is enabled, this register a spill handler for the cached data
+    that deletes the data rather than spilling it. For now, only `RangeIndex`
+    are handled this way. See `SpillManager.register_spill_handler`.
+
+    It is always safe to use this instead of `functools.cached_property`.
+    """
+
+    def __get__(self, instance: T, owner=None):
+        cache_hit = self.attrname in instance.__dict__
+        ret = super().__get__(instance, owner)
+        # Make sure to only register a handler when the cache has changed.
+        # Also, for now, we only handle `RangeIndex` instances that returns
+        # a numerical column with no mask.
+        if (
+            cache_hit
+            or not isinstance(instance, cudf.RangeIndex)
```

Comment:
This is an attempt to address the concerns @shwina points out here and to avoid monkey patching.  

In the original implementation, the decorator was specific to `RangeIndex`, see https://github.com/rapidsai/cudf/pull/12287/commits/8df1adaf57af7285632e4dac075f9d41d0da808f.

After working through this PR a couple of times now, I think having a new decorator specifically for cached Columns is the preferable approach. Would that be acceptable @shwina? 

Either way, I think we should avoid monkey patching. 